File: testbed/synthetic-vuln-java-app/src/main/java/com/example/agent/tools/DangerousTools.java
// Anti-pattern: D003 Java (code-execution tools registered).
// Also fires DF001 + R001.
package com.example.agent.tools;

import dev.langchain4j.agent.tool.Tool;

import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;

public class DangerousTools {

    @Tool("execute a shell command")
    public String shell(String cmd) throws Exception {
        Process p = Runtime.getRuntime().exec(cmd);
        return new String(p.getInputStream().readAllBytes());
    }

    @Tool("spawn a process")
    public void spawn(String binary, String arg) throws Exception {
        new ProcessBuilder(binary, arg).start();
    }

    @Tool("evaluate JavaScript expression")
    public Object evalJs(String code) throws Exception {
        ScriptEngine engine = new ScriptEngineManager().getEngineByName("JavaScript");
        return engine.eval(code);
    }
}

File: testbed/synthetic-vuln-python-app/src/synthetic_vuln_python_app/dangerous_tools.py
"""D003 — code-execution tools registered with the agent."""

import os
import subprocess

from langchain.agents import Tool, initialize_agent
from langchain_community.tools import ShellTool
from langchain_experimental.tools import PythonREPLTool
from langchain.tools import tool
from langchain_openai import ChatOpenAI


def build_dangerous_agent():
    llm = ChatOpenAI()
    tools = [
        PythonREPLTool(),
        ShellTool(),
        Tool(name="exec", func=exec, description="execute python"),
        Tool(name="eval", func=eval, description="evaluate expression"),
    ]
    return initialize_agent(tools, llm)


@tool
def shell_passthrough(cmd: str) -> str:
    """Run a shell command."""
    return subprocess.run(cmd, shell=True, capture_output=True, check=False).stdout.decode()


@tool
def os_system_passthrough(cmd: str) -> int:
    """Run an os.system command."""
    return os.system(cmd)
